astropy/io/ascii/tests/test_qdp.py

```diff
@@ -0,0 +1,135 @@
+import numpy as np
+import pytest
+from astropy.io.ascii import core
+from astropy.io.ascii.qdp import read_table_qdp, write_table_qdp
+from astropy.table import Table, Column, MaskedColumn
+from astropy.utils.exceptions import AstropyUserWarning
+
+
+def test_get_tables_from_qdp_file(tmpdir):
+    example_qdp = """
+    ! Swift/XRT hardness ratio of trigger: XXXX, name: BUBU X-2
+    ! Columns are as labelled
+    READ TERR 1
+    READ SERR 2
+    ! WT -- hard data
+    !MJD            Err (pos)       Err(neg)        Rate            Error
+    53000.123456 2.37847222222222e-05    -2.37847222222222e-05   -0.212439       0.212439
+    55045.099887 1.14467592592593e-05    -1.14467592592593e-05   0.000000        0.000000
+    NO NO NO NO NO
+    ! WT -- soft data
+    !MJD            Err (pos)       Err(neg)        Rate            Error
+    53000.123456 2.37847222222222e-05    -2.37847222222222e-05   0.726155        0.583890
+    55045.099887 1.14467592592593e-05    -1.14467592592593e-05   2.410935        1.393592
+    NO NO NO NO NO
+    ! WT -- hardness ratio
+    !MJD            Err (pos)       Err(neg)        Rate            Error
+    53000.123456 2.37847222222222e-05    -2.37847222222222e-05   -0.292553       -0.374935
+    55045.099887 1.14467592592593e-05    -1.14467592592593e-05   0.000000        -nan
+    """
+
+    path = str(tmpdir.join('test.qdp'))
+
+    with open(path, "w") as fp:
+        print(example_qdp, file=fp)
+
+    table0 = read_table_qdp(fp.name, names=["MJD", "Rate"], table_id=0)
+    assert table0.meta["initial_comments"].lstrip().startswith("Swift")
+    assert table0.meta["comments"].lstrip().startswith("WT -- hard data")
+    table2 = read_table_qdp(fp.name, names=["MJD", "Rate"], table_id=2)
+    assert table2.meta["initial_comments"].lstrip().startswith("Swift")
+    assert table2.meta["comments"].lstrip().startswith("WT -- hardness")
+    assert np.isclose(table2["MJD_nerr"][0], -2.37847222222222e-05)
+
+
+def test_roundtrip(tmpdir):
+    example_qdp = """
```

Comment:
For testing it is nice to include a synthetic file with "simple" values. This makes it easier to verify that it is fully working. For example in my testing of this I used this:
```
example_qdp = """
    ! Initial comment line 1
    ! Initial comment line 2
    READ TERR 1
    READ SERR 3
    ! Table 0 comment
    !MJD            Err (pos)       Err(neg)        Rate            Error
    53000.5   0.25  -0.5   1  1.5  3.5 2
    54000.5   1.25  -1.5   2  2.5  4.5 3
    NO NO NO NO NO
    ! Table 1 comment
    !MJD            Err (pos)       Err(neg)        Rate            Error
    54000.5   2.25  -2.5   3  3.5  5.5 5
    55000.5   3.25  -3.5   4  4.5  6.5 nan
    """

dat = ascii.read(example_qdp, format='qdp', table_id=1, names=['a', 'b', 'c', 'd'])
dat

<Table length=2>
   a     a_perr  a_nerr    b       c     c_err     d   
float64 float64 float64 float64 float64 float64 float64
------- ------- ------- ------- ------- ------- -------
54000.5    2.25    -2.5     3.0     3.5     5.5     5.0
55000.5    3.25    -3.5     4.0     4.5     6.5     nan
```
You can leave this realistic Swift file if you like, but you should include something like my example and then test that every single value in both tables is expected, and that all the columns have the right name and dtype.

Reply:
Should be done